Plutarch/Rational.hs

```diff
@@ -44,120 +47,142 @@ import Plutarch.Builtin (
-import Plutarch.Pair (PPair (..))
-import Plutarch.Show (PShow (pshow'), pshow)
+import Plutarch.Pair (PPair (PPair))
+import Plutarch.Positive (PPositive, ptryPositive)
+import Plutarch.Show (PShow, pshow, pshow')
+import Plutarch.TermCont (tcont, unTermCont)
 import Plutarch.Trace (ptraceError)
-import Plutarch.Unsafe (punsafeCoerce)
+import Plutarch.TryFrom (PTryFrom (PTryFromExcess, ptryFrom'), ptryFrom)
+import Plutarch.Unsafe (punsafeCoerce, punsafeDowncast)
 
 class PFractional (a :: PType) where
   (#/) :: Term s a -> Term s a -> Term s a
   precip :: Term s (a :--> a)
   pfromRational :: Term s (PRational :--> a)
 
 data PRational s
-  = PRational (Term s PInteger) (Term s PInteger)
+  = PRational (Term s PInteger) (Term s PPositive)
   deriving stock (Generic)
-  deriving anyclass (PlutusType)
+  deriving anyclass (PlutusType, PEq)
+
 instance DerivePlutusType PRational where type DPTStrat _ = PlutusTypeScott
 
 instance (PNum a, PFractional a) => Fractional (Term s a) where
   (/) = (#/)
   recip x = precip # x
-  fromRational x = pfromRational #$ pcon $ PRational (pconstant (numerator x)) (pconstant (denominator x))
+  fromRational x =
+    pfromRational #$ pcon $
+      PRational
+        (pconstant $ numerator x)
+        (punsafeDowncast . pconstant $ denominator x)
 
 instance PShow PRational where
   pshow' _ x =
     pshowRat # x
     where
       pshowRat = phoistAcyclic $
         plam $ \n -> pmatch n $ \(PRational x y) ->
-          pshow x <> "/" <> pshow y
+          pshow x <> "/" <> pshow (pto y)
 
 instance PIsData PRational where
-  pfromDataImpl x' = phoistAcyclic (plam $ \x -> pListToRat #$ pmap # pasInt #$ pasList # pforgetData x) # x'
+  pfromDataImpl x' = phoistAcyclic (plam $ \x -> plistToRat #$ pasList # pforgetData x) # x'
+    where
+      plistToRat :: Term s (PBuiltinList PData :--> PRational)
+      plistToRat = plam $ \x ->
+        pcon $
+          PRational (pasInt #$ phead # x)
+            . punsafeDowncast
+            $ pasInt #$ phead #$ ptail # x
   pdataImpl x' =
     phoistAcyclic
-      ( plam $ \x ->
-          punsafeCoerce (pdata $ pRatToList # x :: Term _ (PAsData (PBuiltinList (PAsData PInteger))))
+      ( plam $ \x -> unTermCont $ do
+          PRational a b <- tcont $ pmatch x
+          let res :: Term _ (PBuiltinList (PAsData PInteger))
+              res = pcons # pdata a #$ pcons # pdata (pto b) #$ pnil
+          pure $ pdataImpl res
       )
       # x'
 
-pRatToList :: Term s (PRational :--> PBuiltinList (PAsData PInteger))
-pRatToList = plam $ \x -> pmatch x $ \(PRational a b) ->
-  pcons # pdata a #$ pcons # pdata b #$ punsafeCoerce (pnil :: Term s (PBuiltinList PData))
+newtype Flip f a b = Flip (f b a) deriving stock (Generic)
 
-pListToRat :: Term s (PBuiltinList PInteger :--> PRational)
-pListToRat = plam $ \x -> pcon $ PRational (phead # x) (phead #$ ptail # x)
-
-instance PEq PRational where
```

Comment:
I think we still need this.

Reply:
oh yes certainly, oversight - added back.